### DISCUSSION TITLE
What's wrong with my PICamera?

### DISCUSSION THREAD

**Discussion by JangYongGeum** [community/community#65031] [Other Feature Feedback, Questions, & Ideas]
### Body

I'm connecting picamera v2 to Jetson Nano, and the camera screen pops up like this.
![20230826_215257](https://github.com/community/community/assets/122681467/3bbd4d90-ccad-467e-8566-e5366d2fbd7f)

This is the command I wrote :

gst-launch-1.0 nvarguscamerasrc sensor_mode=0 ! 'video/x-raw(memory:NVMM), width=3820, height=2464, framerate=21/1, format=NV12' ! nvvidconv flip-method=0 ! 'video/x-raw, width=960, height=616' ! nvvidconv ! nvegltransform ! nveglglessink -e

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by trinitronx** [comment]
Whoah!  Looks like some pixel format mis-match going on here.

I'm not familiar with the Jetson Nano and nvidia tools you're using in that command.  However, it looks like it's using NV12 format which is a common pixel format that hardware GPUs use.  You'll probably have to ensure proper conversion to NV12 is happening before handing the data off to the GPU on the Jetson Nano.

Here are some helpful resources:

- [PiCamera docs: Unencoded image capture (YUV format)][1] - Explains some useful inner workings about the PiCamera module's higher-level formats supported by the `picamera` python module.
  - This mentions that it's supporting both:
    - [YUV420](https://en.wikipedia.org/wiki/YUV#Y.E2.80.B2UV420p_.28and_Y.E2.80.B2V12_or_YV12.29_to_RGB888_conversion) (planar)
    - [RGB](https://en.wikipedia.org/wiki/RGB) as interleaved Red/Green/Blue pixel data of equal size
    - Raw [Bayer data](https://picamera.readthedocs.io/en/release-1.13/recipes2.html#bayer-data) (Support for accessing this in `picamera` module has been deprecated, but it's the most low-level raw camera data.)
- [About YUV formats][2] - Compare the `YUV_formats.png` graphic for `I420 / IYUV / YUV420p` with the one for `NV12` format and the charts below
  - [How to convert RGB to NV12 color space using IPP][3] - This S.O. question has a chart showing `NV12` format.  It's also `YUV 4:2:0`... but notice the `U` and `V` planes are **interleaved** rather than sequential in the data frame.
  - [NV12 to YUV444 speed up][4] - More detail on conversion between `NV12` and a _**different**_ type of `YUV 4:4:4` format, but _pay attention to the Illustration picture_.  It depicts the `U` and `V` interleaving of `NV12` format on the left.  Then they convert to `YUV` with equal planes (e.g. `4:4:4`).  It's just a good visual aid when they are talking about "Planes" here.

The banding in your image looks like it's probably due to the mismatch between interleaved `U` and `V` and non-interleaved input from the Raspberry Pi Camera.  You'll need to convert between pixel formats to upload to `NV12` format to the GPU & Nvidia tools.

I'd also suggest checking that your camera is able to get images with [`raspistill`][5] and/or [`picamera`][6] python module first before moving on to the Nvidia GPU side of things.  That's to rule out any hardware defects and that the Picamera is plugged in and operating properly.  The usage of both is pretty simple, as you can see from [this `raspistill`-based project][7] and this [Docker][8]/[`picamera`][9]-based project.

I'm afraid that I can't help much with those `gst-launch` / `nv*` tools, as I'm not familiar with them, nor the Jetson Nano.  The main idea remains the same:  Make sure your input pixel format matches what those tools are expecting.  If they don't match, convert pixel formats.

[1]: https://picamera.readthedocs.io/en/release-1.13/recipes2.html#unencoded-image-capture-yuv-format
[2]: https://gist.github.com/Jim-Bar/3cbba684a71d1a9d468a6711a6eddbeb
[3]: https://stackoverflow.com/questions/37768751/how-to-convert-rgb-to-nv12-color-space-using-ipp
[4]: https://stackoverflow.com/a/67949358/645491
[5]: https://www.raspberrypi.org/app/uploads/2013/07/RaspiCam-Documentation.pdf
[6]: https://picamera.readthedocs.io/en/release-1.13/index.html
[7]: https://github.com/trinitronx/phototimer/blob/master/take.py#L30-L40
[8]: https://github.com/trinitronx/balena-rpi-python-picamera/blob/master/start.sh#L1
[9]: https://github.com/trinitronx/balena-rpi-python-picamera/blob/master/demo.py

**Comment by ghostinhershell** [comment]
Thanks for posting in the GitHub Community, @JangYongGeum! 

We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Projects and Issues category is solely related to conversations around the GitHub products Issues and Projects. I've gone ahead and moved it for you. Good luck!
